### PR TITLE
Error out when any RS has more available pods then its spec replicas

### LINKED ISSUE #29559: [k8s.io] Deployment scaled rollout should not block on annotation check {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/29432/kubernetes-pull-build-test-e2e-gce/50413/

Failed: [k8s.io] Deployment scaled rollout should not block on annotation check {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:88
Expected error:
    <*errors.errorString | 0xc820b929c0>: {
        s: "error waiting for deployment \"nginx\" status to match expectation: total pods available: 17, less than the min required: 18",
    }
    error waiting for deployment "nginx" status to match expectation: total pods available: 17, less than the min required: 18
not to have occurred
```

Happened on a presubmit run in #29432.

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #29559 (hopefully, if not the bot will open new issues for us)

@kubernetes/deployment

### CODE REVIEW

**Comment by pwittrock** [comment]
@janetkuo can you take a look at the failure and also fix the cla?

**Comment by pwittrock** [comment]
@janetkuo It seems like maybe the failed test was supposed to catch the issue that this was meant to fix.  e.g. no scale is expected when observed Pods > desired Replicas

**Comment by janetkuo** [comment]
The failed test has an old RS with 0 replica so it's not likely to be scaled down, so it's a different case to me. I set another `bool` for expect errors since the test case is entering an error state that should be caught by the new error.  WDYT?

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -648,9 +648,16 @@ func filterPodsMatchingReplicaSets(replicaSets []*extensions.ReplicaSet, podList
 		if matchingFunc == nil {
 			continue
 		}
-		rsPods = append(rsPods, podutil.Filter(podList, matchingFunc)...)
+		rsPods := podutil.Filter(podList, matchingFunc)
+		avaPodsCount := countAvailablePods(rsPods, minReadySeconds)
+		if avaPodsCount > rs.Spec.Replicas {
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: Won't this be the case sometimes when a replica set is scaled down?

Comment (bgrant0607) [inline pkg/controller/deployment/util/deployment_util.go]: @janetkuo I agree with @kargakis.

**Comment by kargakis** [comment]
The flake that this was destined to fix is back (https://github.com/kubernetes/kubernetes/issues/30388). I think we should fix `filterPodsMatchingReplicaSets` not to error out when available pods are more than desired since that isn't uncommon when scaling down.

**Comment by janetkuo** [comment]
@kargakis yes you're right

**Pull request merged**